File: frontend/src/types/mission.ts
import type { PointZ } from "./common";
import type {
  CaptureMode,
  ComputationStatus,
  FlightPlanScope,
  InspectionMethod,
  MissionStatus,
  ScanLengthMode,
  ScanRunOrientation,
  ScanWidthSide,
} from "./enums";

export type CameraMode = "AUTO" | "MANUAL";

export type InspectionDirection = "NATURAL" | "REVERSED";

export type MissionDirection = "AUTO" | "NATURAL" | "REVERSED";

export type DjiHeadingMode = "smoothTransition" | "towardPOI" | "followWayline";

export type LhaSelectionMode = "ALL" | "RANGE" | "FROM_THRESHOLD" | "CUSTOM";

export type AngleSource = "PAPI" | "CUSTOM";

export type ThresholdAnchor = "START" | "END";

export interface LhaSelectionRangeParams {
  from: number | null;
  to: number | null;
}

export interface LhaSelectionFromThresholdParams {
  threshold: ThresholdAnchor;
  distance_m: number;
}

export type LhaSelectionRule =
  | { mode: "ALL" }
  | { mode: "RANGE"; params: LhaSelectionRangeParams }
  | { mode: "FROM_THRESHOLD"; params: LhaSelectionFromThresholdParams }
  | { mode: "CUSTOM" };

export type LhaSelectionRules = Record<string, LhaSelectionRule>;

export interface MissionResponse {
  id: string;
  name: string;
  status: MissionStatus;
  airport_id: string;
  created_at: string;
  updated_at: string;
  operator_notes: string | null;
  drone_profile_id: string | null;
  date_time: string | null;
  default_speed: number | null;
  measurement_speed_override: number | null;
  default_altitude_offset: number | null;
  takeoff_coordinate: PointZ | null;
  landing_coordinate: PointZ | null;
  default_capture_mode: CaptureMode | null;
  default_buffer_distance: number | null;
  camera_mode: CameraMode;
  default_white_balance: string | null;
  default_iso: number | null;
  default_shutter_speed: string | null;
  default_focus_mode: "AUTO" | "INFINITY" | null;
  transit_agl: number | null;
  require_perpendicular_runway_crossing: boolean;
  keep_inside_airport_boundary: boolean;
  flight_plan_scope: FlightPlanScope;
  direction: MissionDirection;
  has_unsaved_map_changes: boolean;
  computation_status: ComputationStatus;
  computation_error: string | null;
  computation_started_at: string | null;
  inspection_count: number;
  estimated_duration: number | null;
  supports_geozone_upload?: boolean | null;
  dji_heading_mode?: DjiHeadingMode | null;
}

export interface MissionDetailResponse extends MissionResponse {
  inspections: InspectionResponse[];
}

// surface-scan fields, mirrored once by the ScanConfigFields backend mixin
export interface ScanConfigFields {
  scan_surface_id: string | null;
  scan_length_mode: ScanLengthMode | null;
  scan_length_from: number | null;
  scan_length_to: number | null;
  scan_width: number | null;
  scan_width_side: ScanWidthSide | null;
  scan_height: number | null;
  scan_run_count: number | null;
  scan_run_orientation: ScanRunOrientation | null;
  scan_sidelap_percent: number | null;
}

// runtime key list driving the autosave echo/reset hydration
export const SCAN_FIELDS = [
  "scan_surface_id",
  "scan_length_mode",
  "scan_length_from",
  "scan_length_to",
  "scan_width",
  "scan_width_side",
  "scan_height",
  "scan_run_count",
  "scan_run_orientation",
  "scan_sidelap_percent",
] as const satisfies readonly (keyof ScanConfigFields)[];

export type ScanField = (typeof SCAN_FIELDS)[number];

// shared core mirrored by the template-side InspectionConfigResponse in inspectionTemplate.ts
export interface BaseInspectionConfigFields extends ScanConfigFields {
  altitude_offset: number | null;
  angle_offset_above: number | null;
  angle_offset_below: number | null;
  measurement_speed_override: number | null;
  measurement_density: number | null;
  custom_tolerances: Record<string, number> | null;
  hover_duration: number | null;
  horizontal_distance: number | null;
  sweep_angle: number | null;
  angle_source: AngleSource | null;
  angle_start: number | null;
  angle_end: number | null;
  lha_ids: string[] | null;
  lha_selection_rules: LhaSelectionRules | null;
  capture_mode: CaptureMode | null;
  recording_setup_duration: number | null;
  buffer_distance: number | null;
  height_above_lights: number | null;
  lateral_offset: number | null;
  distance_from_lha: number | null;
  height_above_lha: number | null;
  camera_gimbal_angle: number | null;
  selected_lha_id: string | null;
  lha_setting_angle_override_id: string | null;
  hover_bearing: number | null;
  hover_bearing_reference: "RUNWAY" | "COMPASS" | null;
  descent_start_distance: number | null;
  descent_glide_slope_override: number | null;
  direction: InspectionDirection | null;
  resolved_direction: InspectionDirection | null;
  white_balance: string | null;
  iso: number | null;
  shutter_speed: string | null;
  focus_mode: "AUTO" | "INFINITY" | null;
  optical_zoom: number | null;
}

export interface InspectionConfigResponse extends BaseInspectionConfigFields {
  camera_mode: CameraMode | null;
  camera_preset_id: string | null;
}

export interface InspectionResponse {
  id: string;
  mission_id: string;
  template_id: string;
  config_id: string | null;
  method: InspectionMethod;
  sequence_order: number;
  lha_ids: string[] | null;
  config: InspectionConfigResponse | null;
}

export interface InspectionConfigOverride extends Partial<ScanConfigFields> {
  altitude_offset?: number | null;
  angle_offset_above?: number | null;
  angle_offset_below?: number | null;
  measurement_speed_override?: number | null;
  measurement_density?: number | null;
  custom_tolerances?: Record<string, number> | null;
  hover_duration?: number | null;
  horizontal_distance?: number | null;
  sweep_angle?: number | null;
  angle_source?: AngleSource | null;
  angle_start?: number | null;
  angle_end?: number | null;
  lha_ids?: string[] | null;
  lha_selection_rules?: LhaSelectionRules | null;
  capture_mode?: CaptureMode | null;
  recording_setup_duration?: number | null;
  buffer_distance?: number | null;
  height_above_lights?: number | null;
  lateral_offset?: number | null;
  distance_from_lha?: number | null;
  height_above_lha?: number | null;
  camera_gimbal_angle?: number | null;
  selected_lha_id?: string | null;
  lha_setting_angle_override_id?: string | null;
  hover_bearing?: number | null;
  hover_bearing_reference?: "RUNWAY" | "COMPASS" | null;
  descent_start_distance?: number | null;
  descent_glide_slope_override?: number | null;
  camera_mode?: CameraMode | null;
  direction?: InspectionDirection | null;
  white_balance?: string | null;
  iso?: number | null;
  shutter_speed?: string | null;
  focus_mode?: "AUTO" | "INFINITY" | null;
  optical_zoom?: number | null;
  camera_preset_id?: string | null;
}

export interface MissionCreate {
  name: string;
  airport_id: string;
  operator_notes?: string | null;
  drone_profile_id?: string | null;
  date_time?: string | null;
  default_speed?: number | null;
  measurement_speed_override?: number | null;
  default_altitude_offset?: number | null;
  takeoff_coordinate?: PointZ | null;
  landing_coordinate?: PointZ | null;
  default_capture_mode?: CaptureMode | null;
  default_buffer_distance?: number | null;
  camera_mode?: CameraMode;
  default_white_balance?: string | null;
  default_iso?: number | null;
  default_shutter_speed?: string | null;
  default_focus_mode?: "AUTO" | "INFINITY" | null;
  transit_agl?: number | null;
  require_perpendicular_runway_crossing?: boolean;
  keep_inside_airport_boundary?: boolean;
  flight_plan_scope?: FlightPlanScope;
  direction?: MissionDirection;
  dji_heading_mode?: DjiHeadingMode | null;
}

export interface MissionUpdate {
  name?: string;
  operator_notes?: string | null;
  drone_profile_id?: string | null;
  date_time?: string | null;
  default_speed?: number | null;
  measurement_speed_override?: number | null;
  default_altitude_offset?: number | null;
  takeoff_coordinate?: PointZ | null;
  landing_coordinate?: PointZ | null;
  default_capture_mode?: CaptureMode | null;
  default_buffer_distance?: number | null;
  camera_mode?: CameraMode;
  default_white_balance?: string | null;
  default_iso?: number | null;
  default_shutter_speed?: string | null;
  default_focus_mode?: "AUTO" | "INFINITY" | null;
  transit_agl?: number | null;
  require_perpendicular_runway_crossing?: boolean;
  keep_inside_airport_boundary?: boolean;
  flight_plan_scope?: FlightPlanScope;
  direction?: MissionDirection;
  dji_heading_mode?: DjiHeadingMode | null;
}

export interface InspectionCreate {
  template_id: string;
  method: InspectionMethod;
  config?: InspectionConfigOverride | null;
}

export interface InspectionUpdate {
  method?: InspectionMethod;
  sequence_order?: number;
  config?: InspectionConfigOverride | null;
}

export interface ReorderRequest {
  inspection_ids: string[];
}

export interface ComputationStatusResponse {
  computation_status: ComputationStatus;
  computation_error: string | null;
  computation_started_at: string | null;
}

File: backend/app/schemas/mission.py
"""pydantic schemas for mission and inspection endpoints."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.common import FocusModeStr, ListMeta, WhiteBalanceStr
from app.schemas.geometry import PointZ
from app.schemas.inspection_config import (  # noqa: F401
    AngleSourceStr,
    CameraModeStr,
    CaptureModeStr,
    HoverBearingRefStr,
    InspectionConfigOverride,
    InspectionConfigResponse,
    InspectionDirectionStr,
    LhaSelectionModeStr,
    LhaSelectionRule,
    LhaSelectionRuleAll,
    LhaSelectionRuleCustom,
    LhaSelectionRuleFromThreshold,
    LhaSelectionRuleFromThresholdParams,
    LhaSelectionRuleRange,
    LhaSelectionRuleRangeParams,
    ScanConfigFields,
    ScanLengthModeStr,
    ScanRunOrientationStr,
    ScanWidthSideStr,
    ThresholdAnchorStr,
    _rule_discriminator,
    _validate_transit_altitude,
)

# computation status values - mirrors ComputationStatus enum
ComputationStatusStr = Literal["IDLE", "COMPUTING", "COMPLETED", "FAILED"]

# flight plan scope values - mirrors FlightPlanScope enum
FlightPlanScopeStr = Literal["FULL", "MEASUREMENTS_ONLY"]

# inspection method values - mirrors InspectionMethod enum
InspectionMethodStr = Literal[
    "VERTICAL_PROFILE",
    "HORIZONTAL_RANGE",
    "APPROACH_DESCENT",
    "FLY_OVER",
    "PARALLEL_SIDE_SWEEP",
    "HOVER_POINT_LOCK",
    "MEHT_CHECK",
    "SURFACE_SCAN",
    "RUNWAY_HORIZONTAL_RANGE",
]
# mission-wide direction default - AUTO = solver picks, NATURAL/REVERSED = pin all inspections.
MissionDirectionStr = Literal["AUTO", "NATURAL", "REVERSED"]
# dji wpml heading mode override. smoothTransition (default) interpolates
# body yaw between per-WP angles - works on every documented model and
# depends only on static-angle interpolation, no runtime POI math.
# towardPOI is experimental continuous POI tracking. followWayline is the
# proven fallback that snaps body heading at each waypoint.
DjiHeadingModeStr = Literal["smoothTransition", "towardPOI", "followWayline"]


class InspectionCreate(BaseModel):
    """add inspection to mission."""

    template_id: UUID
    method: InspectionMethodStr
    config: InspectionConfigOverride | None = None


class InspectionUpdate(BaseModel):
    """update inspection within mission."""

    method: InspectionMethodStr | None = None
    config: InspectionConfigOverride | None = None
    sequence_order: int | None = None


class InspectionResponse(BaseModel):
    """inspection response."""

    id: UUID
    mission_id: UUID
    template_id: UUID
    config_id: UUID | None = None
    method: InspectionMethodStr
    sequence_order: int
    lha_ids: list[UUID] | None = None
    config: InspectionConfigResponse | None = None

    model_config = {"from_attributes": True}


class ReorderRequest(BaseModel):
    """reorder inspections by sequence."""

    inspection_ids: list[UUID]


class ReorderResponse(BaseModel):
    """reorder response."""

    reordered: bool


class MissionCreate(BaseModel):
    """create mission."""

    name: str
    airport_id: UUID
    drone_profile_id: UUID | None = None
    operator_notes: str | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = Field(default=None, gt=0)
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = Field(default=None, ge=0)
    camera_mode: CameraModeStr = "AUTO"
    default_white_balance: WhiteBalanceStr | None = None
    default_iso: int | None = Field(default=None, gt=0)
    default_shutter_speed: str | None = Field(default=None, max_length=20)
    default_focus_mode: FocusModeStr | None = None
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool = True
    keep_inside_airport_boundary: bool = True
    flight_plan_scope: FlightPlanScopeStr = "FULL"
    direction: MissionDirectionStr = "AUTO"
    dji_heading_mode: DjiHeadingModeStr | None = None

    @field_validator("transit_agl")
    @classmethod
    def _check_transit_altitude(cls, v: float | None) -> float | None:
        """enforce minimum AGL floor on mission-level cruise altitude."""
        return _validate_transit_altitude(v)


class MissionUpdate(BaseModel):
    """update mission."""

    name: str | None = None
    drone_profile_id: UUID | None = None
    operator_notes: str | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = Field(default=None, gt=0)
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    date_time: datetime | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = Field(default=None, ge=0)
    camera_mode: CameraModeStr | None = None
    default_white_balance: WhiteBalanceStr | None = None
    default_iso: int | None = Field(default=None, gt=0)
    default_shutter_speed: str | None = Field(default=None, max_length=20)
    default_focus_mode: FocusModeStr | None = None
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool | None = None
    keep_inside_airport_boundary: bool | None = None
    flight_plan_scope: FlightPlanScopeStr | None = None
    direction: MissionDirectionStr | None = None
    dji_heading_mode: DjiHeadingModeStr | None = None

    @field_validator("transit_agl")
    @classmethod
    def _check_transit_altitude(cls, v: float | None) -> float | None:
        """enforce minimum AGL floor on mission-level cruise altitude."""
        return _validate_transit_altitude(v)


class MissionResponse(BaseModel):
    """mission response."""

    id: UUID
    name: str
    status: str
    airport_id: UUID
    created_at: datetime
    updated_at: datetime
    operator_notes: str | None = None
    drone_profile_id: UUID | None = None
    date_time: datetime | None = None
    default_speed: float | None = None
    measurement_speed_override: float | None = None
    default_altitude_offset: float | None = None
    takeoff_coordinate: PointZ | None = None
    landing_coordinate: PointZ | None = None
    default_capture_mode: CaptureModeStr | None = None
    default_buffer_distance: float | None = None
    camera_mode: CameraModeStr = "AUTO"
    default_white_balance: WhiteBalanceStr | None = None
    default_iso: int | None = None
    default_shutter_speed: str | None = None
    default_focus_mode: FocusModeStr | None = None
    transit_agl: float | None = None
    require_perpendicular_runway_crossing: bool = True
    keep_inside_airport_boundary: bool = True
    flight_plan_scope: FlightPlanScopeStr = "FULL"
    direction: MissionDirectionStr = "AUTO"
    dji_heading_mode: DjiHeadingModeStr | None = None
    has_unsaved_map_changes: bool = False
    computation_status: ComputationStatusStr = "IDLE"
    computation_error: str | None = None
    computation_started_at: datetime | None = None
    inspection_count: int = 0
    estimated_duration: float | None = None
    # mirrors DroneProfile.supports_geozone_upload from the linked drone profile
    # so the export panel can decide whether to enable the geozone-bundle flag
    # without a second round-trip.
    supports_geozone_upload: bool | None = None

    model_config = {"from_attributes": True}


class MissionDetailResponse(MissionResponse):
    """mission with inspections."""

    inspections: list[InspectionResponse] = []


class ComputationStatusResponse(BaseModel):
    """lightweight computation status for polling."""

    computation_status: ComputationStatusStr
    computation_error: str | None = None
    computation_started_at: datetime | None = None

    model_config = {"from_attributes": True}


class MissionListResponse(BaseModel):
    """mission list response."""

    data: list[MissionResponse]
    meta: ListMeta
